### PR TITLE
Issue #4014: emit PPO parameter metadata

### PR DESCRIPTION
**Pull request by ll7**
Reviewer-critical summary

What changed: `scripts/training/train_ppo.py` now writes a `parameter_summary` block into the existing PPO performance summary JSON and records `policy_parameter_count=...` in training notes when a constructed model is available. This directly advances issue #4014's remaining requirement that PPO / RecurrentPPO-LSTM / PPO-Mamba comparison artifacts include parameter-count metadata alongside throughput.

Why now: closure audit found #4014 is not complete. Merged PR #4475 registered the PPO-Mamba smoke lane, and the latest issue comment says remaining work is matched smoke/full training runs emitting throughput plus parameter-count metadata and diagnostic comparison artifacts. This PR is the smallest nameable progression slice toward that remaining plan.

Claim boundary: this PR proves the PPO training artifact schema can carry parameter-count metadata and that the #4014 PPO-Mamba dry-run emits the field. It does not run the matched training campaign, does not produce the final three-way diagnostic report, and does not establish benchmark or paper-grade evidence.

Required validation: focused tests for parameter-summary serialization/counting and PPO-Mamba dry-run artifact emission, direct PPO-Mamba dry-run command, Ruff check, and final PR readiness.

Remaining blocker: #4014 still needs matched PPO, RecurrentPPO-LSTM, and PPO-Mamba smoke/full training runs with throughput and populated parameter counts, followed by diagnostic comparison artifacts with explicit claim boundary.

Contract delta

New schema fields: existing PPO perf summary JSON now includes `parameter_summary` with `available`, `policy_parameter_count`, `policy_trainable_parameter_count`, `model_parameter_count`, and `model_trainable_parameter_count`.

New fail-closed blockers: none. Dry-runs that intentionally skip model construction record `available: false` with null counts instead of pretending a parameter count exists.

Compatibility impact: existing perf-summary consumers keep all prior keys; this is an additive JSON field. Training behavior and PPO config defaults are unchanged.

Issue: closes no issue; advances https://github.com/ll7/robot_sf_ll7/issues/4014.

Acceptance evidence

| Criterion | Evidence |
| --- | --- |
| Read full issue thread and linked/merged PRs | Read live #4014 issue body/comments and merged PR #4475. Latest maintainer state: #4475 registered PPO-Mamba smoke lane; remaining work is matched training throughput, parameter counts, and diagnostic artifacts. |
| Do not close if acceptance criteria are unmet | #4014 remains open because matched three-way training/report criteria are not met by #4475 or this slice. |
| Smallest remaining slice | Adds parameter-count metadata to the canonical PPO perf summary writer and tests it through the #4014 PPO-Mamba dry-run path. |
| Fragmentation guard | Only one #4014 PR (#4475) found merged in the prior 24h; this PR adds a nameable new capability, not another guardrail/checker packet. |
| Late guidance check | Re-ran `gh issue view 4014 --repo ll7/robot_sf_ll7 --comments --json ...` before PR open; latest comment remains 2026-07-04T18:35:45Z, older than this run. |

Validation

- `uv run ruff check scripts/training/train_ppo.py tests/training/test_train_expert_ppo_contract.py tests/training/test_train_ppo_mamba_config_issue_4014.py` passed.
- `uv run pytest tests/training/test_train_expert_ppo_contract.py::test_write_perf_summary_writes_expected_keys tests/training/test_train_expert_ppo_contract.py::test_model_parameter_summary_counts_policy_parameters tests/training/test_train_ppo_mamba_config_issue_4014.py -q` passed: 5 tests.
- `ROBOT_SF_ARTIFACT_ROOT=$(mktemp -d) uv run python scripts/training/train_ppo.py --config configs/training/ppo/issue_4014_ppo_mamba_smoke.yaml --dry-run --log-level WARNING` passed; perf JSON included `parameter_summary.available=false` for dry-run.
- `BASE_REF=origin/main scripts/dev/pr_ready_check.sh` passed on dirty tree before commit; log: `/tmp/issue4014_pr_ready.log`.
- `PR_READY_PR_BODY_FILE=/tmp/issue4014_pr_body.md PR_READY_MODE=final BASE_REF=origin/main scripts/dev/pr_ready_check.sh` passed before PR open.

Out of scope

No full benchmark campaign run, no Slurm/GPU submission, no paper/dissertation claim edits.

<details>
<summary>Governance and artifact notes</summary>

State propagation: issue comments were not posted because `comment_issue_or_pr` is false. No tracked issue-state file was added because this PR is a schema/behavior slice and the request forbids encoding transient queue-routing state in tracked files; remaining work is captured in this PR body.

Generated artifacts: local readiness/coverage outputs under `output/` and direct dry-run artifacts under `/tmp/issue4014-dryrun.*` are disposable validation artifacts and are not committed.

Reviewer handoff: Claude Code on `imech156-u` owns full PR gate, improvement cycle, and merge authority after this PR opens.

</details>

### CODE REVIEW

**Comment by ll7** [comment]
## Gate review (Opus 4.8 review gate) — head `96b29201a` — standard lens (diagnostic training-artifact schema)

**Verdict: no blocking findings — merge on green.** Low-risk additive slice for #4014 (+129/-0, 3 files: `scripts/training/train_ppo.py` +45 plus two test files). Adds a `parameter_summary` block (`available` + policy/model total & trainable counts) to the existing PPO perf-summary JSON and a `policy_parameter_count=...` training note. No planner/wrapper/metric/scenario semantics; training behavior and PPO defaults unchanged. Body: "advances #4014; closes no issue" — correct, #4014 stays open.

### Open review-bot threads (exhaustive enumeration)
- **0 open review threads.** GraphQL `reviewThreads` empty; CodeRabbit "Review completed", Gemini rate-limited (advisory only). Nothing to fix or waive.

### Findings from the diff
1. **[VERIFIED — fail-honest]** `_model_parameter_summary(None)` returns `available: False` with all counts `null` rather than fabricating a zero/placeholder count; `_count_parameters` returns `None` when the module exposes no callable `parameters()`, and skips params lacking a callable `numel()`. A dry-run (no constructed model) therefore records the contract without pretending a count exists — matched by `test_ppo_mamba_dry_run_emits_parameter_summary_placeholder`.
2. **[LOW — non-blocking, by design]** For a real SB3 PPO/RecurrentPPO model, `model.parameters` is generally not a callable module method (the `nn.Module` is `model.policy`), so `model_parameter_count`/`model_trainable_parameter_count` will typically serialize as `null` while `policy_parameter_count`/`policy_trainable_parameter_count` populate. This is honest (null, not fake) and the #4014 requirement is the policy parameter count, which works. Worth noting for the eventual three-way diagnostic report so the model-level nulls aren't misread as "0 params". Tracked on open parent #4014.
3. **[VERIFIED — additive/back-compat]** `parameter_summary` is an additive JSON key; the `_write_perf_summary` default falls back to the unavailable summary, so existing perf-summary consumers keep all prior keys.
4. **[VERIFIED — test-path hygiene]** `REPO_ROOT`/`CONFIG_PATH` are `Path(__file__).resolve().parents[2]`-anchored (not cwd-relative); the dry-run subprocess runs with `cwd=REPO_ROOT` and `ROBOT_SF_ARTIFACT_ROOT=tmp_path`. Counting logic (total vs trainable) directly unit-tested.

No in-scope defects to fix. Merging once the pending checklist (fast-feedback / smoke / pr-body-contracts) passes green.
